nixos/nextcloud: fix upgrade path from 19.09 to 20.03

Ma27 opened the pull request:
###### Motivation for this change

As described in #82056, the upgrade path of `nextcloud` from 19.09 to 20.03 is broken as we'd have to move two major versions forward which isn't supported by Nextcloud.

This PR introduces versioned attributes for Nextcloud and a few assertions (relying on the `stateVersion`) to help users with upgrading.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

dasJ commented on pkgs/servers/nextcloud/intermediate.nix:
```diff
@@ -0,0 +1,25 @@
+
+let
+  override = nextcloud.overrideAttrs (old: rec {
+    version = "17.0.2";
```
Umm, do we have a way to let the ryantm bot ignore this?

Ma27 replied:
Good point.
cc @ryantm in case this is getting merged, could you please blacklist this package for r-ryantm?

ryantm replied:
Sure, I can blacklist the top-level nextcloud attrpath from being updated. Would that cover it for you?

Ma27 replied:
The only thing to blacklist would be `nextcloud-intermediate`, but that should be sufficient - thanks!. Please wait with that change until this PR is merged though, this is basically a suggestion how to fix the overall issue and things in here are subject to change.

aanderse commented:
@Ma27 does it make sense to just call the packages by their version number? We could default anything with an old `stateVersion` to `nextcloud17` like you suggested, and maybe add a `warnings` to encourage users to change `.package` to `nextcloud18` **after** upgrading to `nextcloud17`? If we entirely remove the `nextcloud` package and only provide `nextcloudXX`, keeping `warnings` (based on `.package`) up to date I think that might be less confusing. :thinking:

Ma27 commented:
Hmm, that's an even better idea :+1: 

Will update the patches soon :)

ryantm commented:
If you do it this way, you won't need to worry about nixpkgs-update, because it makes sure the version number is compatible with the attrpath: https://github.com/ryantm/nixpkgs-update/blob/master/src/Version.hs#L35

bachp commented:
What about keeping the `nextcloud` package, but but making it depend on the state version? This would give a default for people starting with a new installation.

fpletz commented:
I think @aanderse is on the right track here. After discussing this with @Ma27, we found some edgy use cases we should look out for:

  1. Users who are currently overriding `nixpkgs.nextcloud` with a custom derivation or are overriding the existing one need a solution when the `nextcloud` attribute is removed and not used in the NixOS module anymore
      * They would otherwise silently receive our default in the NixOS module instead of their overridden package
      * Possible solutions if `nixpkgs.nextcloud` is to be removed:
        1. Define `nixpkgs.nextcloud` to throw an exception which would tell users overriding this derivation that it is was removed. In the NixOS module, check if `nixpkgs.nextcloud` evaluates. If it does, use it, else use our default logic.
        2. Remove `nixpkgs.nextcloud` and add an assertion in the module. If `nixpkgs.nextcloud` is defined but `services.nextcloud.package` isn't, an error is printed that tells the user to use `services.nextcloud.package` to override the Nextcloud package used in the module.
      * While the first solution would nicer because it's easier grasp what's going on without the module, it still encourages using `nixpkgs.nextcloud` to override the package while a better solution exists. In the implementation, we would have to resort to `tryEval` which would also catch other causes of evaluation errors, i.e. using `nixpkgs.nextcloud.overrideAttrs` would also be catched and thus the logic would silently assume that this user wants our default even though he has clearly used an override.
      * The latter solution explicitly tells all users who are using overrides and the NixOS module, that they should override the package via the new `services.nextcloud.package` option. We suggest to go this route.

  2. We should only ship Nextcloud releases that are supported by upstream in our stable releases and master. People who are not updating often (for whatever reason) might skip an intermediate version.
      * Choosing an intermediate Nextcloud release after a new NixOS release based on the `stateVersion` only works for one intermediate release reliably when users rarely upgrade
      * We should mention in the documentation that users of the Nextcloud module are supposed to update their to recent (channel) releases i.e. every month so every major version bump reaches them

So to sum up, we would suggest do the following:

* Remove `nixpkgs.nextcloud` and introduce Nextcloud attributes for every supported major branch like `nextcloud18`
* Introduce `stateVersion` dependent default for `services.nextcloud.package`
* For case 1: Introduce an assertion as per solution b to inform all users overriding the package to use the module option instead.
* For case 2: Mention in the documentation that we backport major releases to stable channels and users are supposed to update regularly.

fpletz commented:
@bachp `stateVersion` is a NixOS concept/option which should IMHO not influence anything attribute in nixpkgs. NixOS configurations and their options depend on nixpkgs but we AFAIK have no precedent for the other direction (other than the explicit `nixpkgs.*` space in NixOS of course).

Ma27 committed:
nixos/nextcloud: implement a safe upgrade-path between 19.09 and 20.03

It's impossible to move two major-versions forward when upgrading
Nextcloud. This is an issue when comming from 19.09 (using Nextcloud 16)
and trying to upgrade to 20.03 (using Nextcloud 18 by default).

This patch implements the measurements discussed in #82056 and #82353 to
improve the update process and to circumvent similar issues in the
future:

* `pkgs.nextcloud` has been removed in favor of versioned attributes
  (currently `pkgs.nextcloud17` and `pkgs.nextcloud18`). With that
  approach we can safely backport major-releases in the future to
  simplify those upgrade-paths and we can select one of the
  major-releases as default depending on the configuration (helpful to
  decide whether e.g. `pkgs.nextcloud17` or `pkgs.nextcloud18` should be
  used on 20.03 and `master` atm).

* If `system.stateVersion` is older than `20.03`, `nextcloud17` will be
  used (which is one major-release behind v16 from 19.09). When using a
  package older than the latest major-release available (currently v18),
  the evaluation will cause a warning which describes the issue and
  suggests next steps.

  To make those package-selections easier, a new option to define the
  package to be used for the service (namely
  `services.nextcloud.package`) was introduced.

* If `pkgs.nextcloud` exists (e.g. due to an overlay which was used to
  provide more recent Nextcloud versions on older NixOS-releases), an
  evaluation error will be thrown by default: this is to make sure that
  `services.nextcloud.package` doesn't use an older version by accident
  after checking the state-version. If `pkgs.nextcloud` is added
  manually, it needs to be declared explicitly in
  `services.nextcloud.package`.

* The `nixos/nextcloud`-documentation contains a
  "Maintainer information"-chapter  which describes how to roll out new
  Nextcloud releases and how to deal with old (and probably unsafe)
  versions.

Closes #82056

Ma27 commented:
@GrahamcOfBorg test nextcloud

mmilata commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -64,6 +66,10 @@ in {
+    package = mkOption {
+      type = types.package;
+      description = "Which package to use for the Nextcloud instance.";
+    };
```
```suggestion
    relatedPackages = [ "nextcloud17" "nextcloud18" ];
    };
```
RHS can probably be extracted from `pkgs`.

Ma27 commented:
Rebased to resolve the merge-conflict in the release-notes.

@fpletz @bachp @aanderse @dasJ would it be possible to get another review from you? :)

aanderse commented:
@Ma27 quick glance over it and looks fine :+1:

Ma27 commented:
@worldofpeace @disassembler would you mind taking a look at this? This should IMHO land in 20.03 :)

worldofpeace commented on pkgs/servers/nextcloud/default.nix:
```diff
@@ -1,24 +1,37 @@
+      maintainers = with maintainers; [ schneefux bachp globin fpletz ma27 ];
+      license = licenses.agpl3Plus;
+      platforms = with platforms; unix;
+      knownVulnerabilities = optional insecure "Nextcloud version ${version} is EOL!";
```
```suggestion
      knownVulnerabilities = optional insecure "Nextcloud version ${version} is EOL";
```

We don't have to shout :smile:

Ma27 replied:
Fixed :+1:

worldofpeace commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -309,10 +316,31 @@ in {
+
+      services.nextcloud.package = with pkgs;
+        mkDefault (
+          if pkgs ? nextcloud
```
`pkgs ? nextcloud` will always be false because it's removed.  Should this be `cfg.package`?

Ma27 replied:
It will be false unless it's defined in an overlay. If that's the case, an assertion will be thrown. This is done to make sure that no downgrade happens (e.g. Nextcloud 18.x in the overlay, but Nextcloud 17.x selected for `nextcloud.package`). If one wants to ues `pkgs.nextcloud` (from an overlay), `services.nextcloud.package` should be set to `pkgs.nextcloud` explicitly.

worldofpeace commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5356,7 +5356,8 @@ in
 
   grocy = callPackage ../servers/grocy { };
 
-  nextcloud = callPackage ../servers/nextcloud { };
```
Should we do a `throw` for migration also?

Ma27 replied:
In that case, `pkgs ? nextcloud` would be true, right? (for further context about this, I recommend to read https://github.com/NixOS/nixpkgs/pull/82353#issuecomment-598269471 since @fpletz and I already had a fairly long discussion about possible corner-cases :D)

worldofpeace replied:
Awesome

worldofpeace reviewed (approved):
I approve of the backport to 20.03

Ma27 commented:
Great, thanks a lot for the review! Will merge tomorrow then :)

Ma27 commented:
Backported to 20.03 as d148bb00cf642eb6c1ae8a75b7dd7c303d8b7090. Thanks a lot to everyone involved!
